File: java/03-advanced/05-data-structures/LinkedListImpl.java
/**
 * LESSON: Custom Linked List Implementation
 * A singly linked list built from scratch to understand the data structure.
 * Each node holds data and a reference to the next node.
 *
 *   [data|next] -> [data|next] -> [data|next] -> null
 */
public class LinkedListImpl {
    public static void main(String[] args) {
        CustomLinkedList list = new CustomLinkedList();

        // Adding elements
        System.out.println("--- Adding Elements ---");
        list.addLast(10);
        list.addLast(20);
        list.addLast(30);
        list.addFirst(5);
        list.addLast(40);
        list.display(); // 5 -> 10 -> 20 -> 30 -> 40

        System.out.println("Size: " + list.size());

        // Accessing
        System.out.println("\n--- Accessing ---");
        System.out.println("Get index 0: " + list.get(0));
        System.out.println("Get index 2: " + list.get(2));
        System.out.println("Contains 20? " + list.contains(20));
        System.out.println("Contains 99? " + list.contains(99));

        // Removing
        System.out.println("\n--- Removing ---");
        list.removeFirst();
        list.display(); // 10 -> 20 -> 30 -> 40
        list.removeLast();
        list.display(); // 10 -> 20 -> 30
        list.removeValue(20);
        list.display(); // 10 -> 30

        // Rebuilding for reverse demo
        list.addFirst(5);
        list.addLast(40);
        list.addLast(50);
        System.out.println("\n--- Before Reverse ---");
        list.display(); // 5 -> 10 -> 30 -> 40 -> 50

        list.reverse();
        System.out.println("--- After Reverse ---");
        list.display(); // 50 -> 40 -> 30 -> 10 -> 5
    }
}

class CustomLinkedList {
    // Inner Node class
    private class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node head;
    private int size;

    CustomLinkedList() {
        head = null;
        size = 0;
    }

    // Add to beginning - O(1)
    void addFirst(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
        size++;
    }

    // Add to end - O(n)
    void addLast(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
        size++;
    }

    // Remove first - O(1)
    void removeFirst() {
        if (head == null) throw new RuntimeException("List is empty");
        head = head.next;
        size--;
    }

    // Remove last - O(n)
    void removeLast() {
        if (head == null) throw new RuntimeException("List is empty");
        if (head.next == null) {
            head = null;
        } else {
            Node current = head;
            while (current.next.next != null) {
                current = current.next;
            }
            current.next = null;
        }
        size--;
    }

    // Remove by value - O(n)
    boolean removeValue(int data) {
        if (head == null) return false;
        if (head.data == data) { removeFirst(); return true; }

        Node current = head;
        while (current.next != null) {
            if (current.next.data == data) {
                current.next = current.next.next;
                size--;
                return true;
            }
            current = current.next;
        }
        return false;
    }

    // Get by index - O(n)
    int get(int index) {
        if (index < 0 || index >= size) throw new IndexOutOfBoundsException("Index: " + index);
        Node current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.data;
    }

    // Search - O(n)
    boolean contains(int data) {
        Node current = head;
        while (current != null) {
            if (current.data == data) return true;
            current = current.next;
        }
        return false;
    }

    // Reverse the list - O(n)
    void reverse() {
        Node prev = null;
        Node current = head;
        while (current != null) {
            Node next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    int size() { return size; }

    void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data);
            if (current.next != null) System.out.print(" -> ");
            current = current.next;
        }
        System.out.println();
    }
}

File: cpp-practice/03-advanced/04-data-structures/linked_list_impl.cpp
/**
 * LESSON: Custom Linked List Implementation
 * Singly linked list built from scratch with templates.
 */
#include <iostream>
#include <stdexcept>

template <typename T>
class LinkedList {
    struct Node {
        T data;
        Node* next;
        Node(const T& d) : data(d), next(nullptr) {}
    };

    Node* head = nullptr;
    int count = 0;

public:
    ~LinkedList() {
        while (head) {
            Node* temp = head;
            head = head->next;
            delete temp;
        }
    }

    void addFirst(const T& data) {
        Node* newNode = new Node(data);
        newNode->next = head;
        head = newNode;
        count++;
    }

    void addLast(const T& data) {
        Node* newNode = new Node(data);
        if (!head) { head = newNode; }
        else {
            Node* curr = head;
            while (curr->next) curr = curr->next;
            curr->next = newNode;
        }
        count++;
    }

    void removeFirst() {
        if (!head) throw std::runtime_error("List is empty");
        Node* temp = head;
        head = head->next;
        delete temp;
        count--;
    }

    void removeLast() {
        if (!head) throw std::runtime_error("List is empty");
        if (!head->next) { delete head; head = nullptr; }
        else {
            Node* curr = head;
            while (curr->next->next) curr = curr->next;
            delete curr->next;
            curr->next = nullptr;
        }
        count--;
    }

    bool contains(const T& data) const {
        Node* curr = head;
        while (curr) {
            if (curr->data == data) return true;
            curr = curr->next;
        }
        return false;
    }

    T get(int index) const {
        if (index < 0 || index >= count) throw std::out_of_range("Index out of range");
        Node* curr = head;
        for (int i = 0; i < index; i++) curr = curr->next;
        return curr->data;
    }

    void reverse() {
        Node* prev = nullptr;
        Node* curr = head;
        while (curr) {
            Node* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        head = prev;
    }

    int size() const { return count; }
    bool empty() const { return head == nullptr; }

    void display() const {
        Node* curr = head;
        while (curr) {
            std::cout << curr->data;
            if (curr->next) std::cout << " -> ";
            curr = curr->next;
        }
        std::cout << "\n";
    }
};

int main() {
    LinkedList<int> list;

    std::cout << "--- Adding ---\n";
    list.addLast(10);
    list.addLast(20);
    list.addLast(30);
    list.addFirst(5);
    list.addLast(40);
    list.display(); // 5 -> 10 -> 20 -> 30 -> 40

    std::cout << "\n--- Accessing ---\n";
    std::cout << "Size: " << list.size() << "\n";
    std::cout << "get(0): " << list.get(0) << "\n";
    std::cout << "get(2): " << list.get(2) << "\n";
    std::cout << "Contains 20? " << std::boolalpha << list.contains(20) << "\n";

    std::cout << "\n--- Removing ---\n";
    list.removeFirst();
    list.display();
    list.removeLast();
    list.display();

    std::cout << "\n--- Reverse ---\n";
    list.addFirst(5);
    list.addLast(40);
    list.display();
    list.reverse();
    list.display();

    // String list
    std::cout << "\n--- String List ---\n";
    LinkedList<std::string> names;
    names.addLast("Alice");
    names.addLast("Bob");
    names.addLast("Charlie");
    names.display();

    return 0;
}
